Support custom resource types being added to API_PROFILES

derekbekoe opened the pull request:
- Only list through RTs in the current profile instead of all (no need to go through all)

Marked as DNM for now to get feedback and also, it includes a change to storage that we wouldn't want to actually merge.

williexu commented on src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/__init__.py:
```diff
@@ -4,19 +4,24 @@
+        custom_resource_types = {'latest': {CUSTOM_MGMT_STORAGE: '2015-06-15'}}
         super(StorageCommandsLoader, self).__init__(cli_ctx=cli_ctx,
                                                     resource_type=ResourceType.DATA_STORAGE,
+                                                    custom_resource_types=custom_resource_types,
```
Kind of strange adding it here. The other kwargs are more sort of defaults that the commands from this loader will use.
The `custom_resource_types` kwargs is used to inject new resourcetypes to the CLI. 
Calling a function from core might make more sense.

derekbekoe replied:
Moved to calling a function.

williexu commented on src/azure-cli-core/azure/cli/core/__init__.py:
```diff
@@ -289,6 +289,11 @@ def __init__(self, cli_ctx=None, min_profile=None, max_profile='latest',
+            from azure.cli.core.profiles import API_PROFILES
+            for profile, v in custom_resource_types.items():
+                if profile in API_PROFILES:
+                    API_PROFILES[profile].update(v)
```
Instead of here, how about adding this update as a callable function in azure.cli.core.profiles?

derekbekoe replied:
I've gone with that approach.

tjprescott commented on src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/__init__.py:
```diff
@@ -4,19 +4,24 @@
         from azure.cli.core.commands import CliCommandType
 
         storage_custom = CliCommandType(operations_tmpl='azure.cli.command_modules.storage.custom#{}')
+        custom_resource_types = {'latest': {CUSTOM_MGMT_STORAGE: '2015-06-15'}}
```
This structure isn't really intuitive.  It should be validated in the __init__. Otherwise an error will likely result in a stack trace.

derekbekoe replied:
This will be resolved with Willie's suggestion.

derekbekoe commented:
Addressed feedback, please take another look.

Usage would be something like this:

```
class StorageCommandsLoader(AzCommandsLoader):
     def __init__(self, cli_ctx=None):
         CUSTOM_MGMT_STORAGE = CustomResourceType('azure.mgmt.storage', 'StorageManagementClient')
        augment_api_profiles('latest', CUSTOM_MGMT_STORAGE, '2015-06-15')
```

tjprescott commented on src/azure-cli-core/azure/cli/core/profiles/__init__.py:
```diff
@@ -89,3 +89,15 @@ def get_sdk(cli_ctx, resource_type, *attr_args, **kwargs):
 }
+
+
+def augment_api_profiles(profile_name, resource_type, api_version):
```
So if I wanted to add a resource type to multiple profiles I would have:
```
augment_api_profiles('latest', MY_RESOURCE_TYPE, '2016-05-01')
augment_api_profiles('2017-01-01-profile', MY_RESOURCE_TYPE, '2016-05-01')
...
```

A more concise syntax to consider:
```
register_resource_type(MY_RESOURCE_TYPE, {
  'latest': '2016-05-01',
  '2017-01-01-profile': '2016-05-01'
})
```
Essentially you specify the resource type once and then a dictionary of `profile: version` pairs.

williexu replied:
Requiring a dictionary argument seems confusing.

derekbekoe replied:
I agree with @williexu, the dict format seems confusing and that was an original comment in my first implementation that allowed users to pass in a dict.

I could potentially change the method name from `augment_api_profiles` to `register_resource_type` if that's better for everyone.

williexu replied:
yes to `register_resource_type`

tjprescott replied:
I didn't see anything wrong with the dictionary implementation in the original. I don't like that I would have to repeat so much code repeatedly calling "augment_api_profiles" (or "register_resource_type") once for each resource type and each profile.  I'd be curious what @troydai thinks here. 

Something that looks closest to the API profile structure seems most logical to me (I have a feeling that's probably what your original implementation looked like--but I didn't have problems with that).

```Python
register_resource_types({
  'latest': {
   MY_RT_1: '2018-01-01',
   MY_RT_2: '2018-01-01'
  }
  '2018-01-01-profile': {
   MY_RT_1: '2018-01-01',
   MY_RT_2: '2018-01-01'
  }
})
```

It's essentially just a profile dictionary merge. I don't think that's necessarily confusing--just would just need to validate it.

williexu replied:
@tjprescott  why not:
```
for  profile in 'latest', '2018-...':
    for rt in RT_1, RT_2:
          register_resource_type(profile, rt, version)
```

tjprescott commented on src/azure-cli-core/azure/cli/core/profiles/_shared.py:
```diff
@@ -24,6 +24,12 @@ def __str__(self):
 PROFILE_TYPE = object()
 
 
+class CustomResourceType(object):  # pylint: disable=too-few-public-methods
```
Also, what if you have to add an endpoint type in addition to or instead of a resource type?  How would that be supported?

derekbekoe replied:
That would be part of https://github.com/Azure/azure-cli/issues/4786 which this does not address.

tjprescott replied:
Would it not make sense to implement these both at the same time?

derekbekoe commented:
While Profiles and Endpoints for clouds are related, they are intentionally separated in the source code because you can use the same profile with multiple clouds and there's no relation between the two in terms of Azure.

That's why the cloud and profile code lives separately:
https://github.com/Azure/azure-cli/blob/dev/src/azure-cli-core/azure/cli/core/cloud.py
https://github.com/Azure/azure-cli/blob/dev/src/azure-cli-core/azure/cli/core/profiles/__init__.py

I can take a look at adding support for that next though...

troydai commented on src/azure-cli-core/azure/cli/core/__init__.py:
```diff
@@ -449,10 +449,11 @@ def get_op_handler(self, operation):
         from azure.cli.core.profiles._shared import get_versioned_sdk_path
 
-        for rt in ResourceType:
+        resource_types_in_profile = AZURE_API_PROFILES[self.cli_ctx.cloud.profile].keys()
```
if `AZURE_API_PROFILE` is a map, you don't need the `.keys()` here. Iterate a map is iterating on its keys.

```
>>> m={'a':1,'b':2}
>>> for i in m:
...   print(i)
... 
a
b
```

troydai replied:
Any chance a `KeyError` throws here?
